### PR TITLE
Fix deprecated usage of `null` on `string` params of internal functions

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
Related to #4883

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [x] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Database/Postgre/Connection.php [hunk 1]
```diff
@@ -433,8 +433,11 @@ protected function buildDSN()
             $this->DSN = "host={$this->hostname} ";
         }
 
-        if (! empty($this->port) && ctype_digit($this->port)) {
-            $this->DSN .= "port={$this->port} ";
+        // ctype_digit only accepts strings
+        $port = (string) $this->port;
+
+        if (! empty($port) && ctype_digit($port)) {
```
Comment (MGatner) [inline system/Database/Postgre/Connection.php:439]:
```suggestion
        if ($port !== '' && ctype_digit($port)) {
```
Not sure if `'0'` is significant here?

File: system/Validation/Rules.php [hunk 2]
```diff
@@ -91,20 +80,15 @@ public function greater_than_equal_to(?string $str, string $min): bool
...
      *    is_not_unique[menu.id,active,1]
-     *
-     * @param string $str
      */
     public function is_not_unique(?string $str, string $field, array $data): bool
     {
-        // Grab any data for exclusion of a single row.
         [$field, $whereField, $whereValue] = array_pad(explode(',', $field), 3, null);
 
-        // Break the table and field apart
```
Comment (MGatner) [inline system/Validation/Rules.php:102]: I find these comments helpful.

**Review by MGatner — CHANGES_REQUESTED** [review]
Only one I didn't look very closely at is `Time`, I can revisit if you get two other reviews.

File: system/Validation/FormatRules.php [hunk 3]
```diff
@@ -240,17 +245,14 @@ public function valid_emails(?string $str = null): bool
 
     /**
      * Validate an IP address (human readable format or binary string - inet_pton)
-     *
-     * @param string $ip    IP Address
-     * @param string $which IP protocol: 'ipv4' or 'ipv6'
```
Comment (kenjis) [inline system/Validation/FormatRules.php:245]: This phpdoc is helpful.

File: system/Validation/Rules.php [hunk 4]
```diff
@@ -183,9 +158,6 @@ public function less_than_equal_to(?string $str, string $max): bool
 
     /**
      * Matches the value of another field in $data.
-     *
-     * @param string $str
-     * @param array  $data Other field/value pairs
```
Comment (kenjis) [inline system/Validation/Rules.php:188]: This phpdoc is helpful.

File: system/Validation/Validation.php [hunk 5]
```diff
@@ -647,26 +638,19 @@ public function setError(string $field, string $error): ValidationInterface
...
      */
     protected function getErrorMessage(string $rule, string $field, ?string $label = null, ?string $param = null, ?string $value = null): string
     {
-        // Check if custom message has been defined by user
         if (isset($this->customErrors[$field][$rule])) {
             $message = lang($this->customErrors[$field][$rule]);
         } else {
-            // Try to grab a localized version of the message...
-            // lang() will return the rule name back if not found,
-            // so there will always be a string being returned.
```
Comment (kenjis) [inline system/Validation/Validation.php:662]: This comment is helpful.

File: system/Validation/Validation.php [hunk 6]
```diff
@@ -617,17 +613,12 @@ public function getError(?string $field = null): string
...
-     * Excluded from code coverage because that it always run as cli
+     * @return array<string, string>
      *
      * @codeCoverageIgnore
      */
     public function getErrors(): array
     {
-        // If we already have errors, we'll use those.
-        // If we don't, check the session to see if any were
-        // passed along from a redirect_with_input request.
```
Comment (kenjis) [inline system/Validation/Validation.php:630]: This comment is helpful.

File: system/Validation/Rules.php [hunk 7]
```diff
@@ -270,20 +234,13 @@ public function required_with($str = null, ?string $fields = null, array $data =
             throw new InvalidArgumentException('You must supply the parameters: fields, data.');
         }
 
-        $fields = explode(',', $fields);
-
-        // If the field is present we can safely assume that
-        // the field is here, no matter whether the corresponding
-        // search field is present or not.
```
Comment (kenjis) [inline system/Validation/Rules.php:277]: This comment is helpful.

File: system/Validation/Rules.php [hunk 8]
```diff
@@ -270,20 +234,13 @@ public function required_with($str = null, ?string $fields = null, array $data =
...
+        $fields  = explode(',', $fields);
         $present = $this->required($str ?? '');
 
         if ($present) {
             return true;
         }
 
-        // Still here? Then we fail this test if
-        // any of the fields are present in $data
-        // as $fields is the lis
```
Comment (kenjis) [inline system/Validation/Rules.php:286]: This comment is helpful.

File: system/Validation/Rules.php [hunk 9]
```diff
@@ -311,19 +268,13 @@ public function required_without($str = null, ?string $fields = null, array $dat
             throw new InvalidArgumentException('You must supply the parameters: fields, data.');
         }
 
-        $fields = explode(',', $fields);
-
-        // If the field is present we can safely assume that
-        // the field is here, no matter whether the corresponding
-        // search field is present or not.
```
Comment (kenjis) [inline system/Validation/Rules.php:318]: This comment is helpful.

File: system/Validation/Rules.php [hunk 10]
```diff
@@ -311,19 +268,13 @@ public function required_without($str = null, ?string $fields = null, array $dat
...
+        $fields  = explode(',', $fields);
         $present = $this->required($str ?? '');
 
         if ($present) {
             return true;
         }
 
-        // Still here? Then we fail this test if
-        // any of the fields are not present in $data
```
Comment (kenjis) [inline system/Validation/Rules.php:326]: This comment is helpful.

**Review by kenjis — APPROVED** [review]

**Review by MGatner — APPROVED** [review]

**Pull request merged**